File: Templates/dijkstra-e-log-e.cpp
vector<long long> dijk(const vector<vector<array<int, 2>>>& adj, int node) {
    vector<long long> dist(adj.size(), 1e18); dist[node] = 0;
    vector<bool> visited(adj.size(), false);
    priority_queue<array<long long, 2>, vector<array<long long, 2>>, greater<array<long long, 2>>> pq;
    // pq's elements consist of (distance, node)
    pq.push({0, node});
    while (!pq.empty()) {
        long long process = pq.top()[1]; pq.pop();
        if (visited[process]) continue;
        visited[process] = true;
        for (array<int, 2> i : adj[process]) {
            // i = (node, weight to node)
            dist[i[0]] = min(dist[i[0]], dist[process] + i[1]);
            pq.push({dist[i[0]], i[0]});
        }
    }
    return dist;
}

File: Templates/dijkstra-e-log-e.java
static class Pair<K, V> {
    public K f; public V s;
    public Pair(K f, V s) {
        this.f = f; this.s = s;
    }
}
static long[] dijk(ArrayList<Pair<Integer, Integer>>[] adj, int node) {
    long[] dist = new long[adj.length];
    Arrays.fill(dist, 1000000000000000000L); dist[node] = 0;
    PriorityQueue<Pair<Long, Integer>> pq = new PriorityQueue<>(Comparator.comparingLong(x -> x.f));
    pq.add(new Pair<>(0L, node));
    while (!pq.isEmpty()) {
        Pair<Long, Integer> process = pq.remove();
        if (process.f > dist[process.s]) continue;
        for (Pair<Integer, Integer> i : adj[process.s]) {
            if (dist[process.s] + i.s < dist[i.f]) {
                dist[i.f] = dist[process.s] + i.s;
                pq.add(new Pair<>(dist[i.f], i.f));
            }
        }
    }
    return dist;
}
